### PR TITLE
fix(datastore): ignore sqlite exceptions during cascading delete

### PR DESCRIPTION
**Pull request by raphkim**
*Issue #, if available:*

*Description of changes:*
Ignore `SQLiteException` during cascading delete for general stability. (local deletions will still be carried out, but remote deletions are cut short)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by richardmcclellan** [comment]
This change looks good!  Could you also add a unit test for the scenario that this fixes?

**Review by jamesonwilliams — COMMENTED** [review]
> Ignore SQLiteException during cascading delete for general stability. (local deletions will still be carried out, but remote deletions are cut short)

Can you expand upon this a bit more in the PR description? Is it not possible to try enqueuing the remote deletion, even if the local one fails?

Also, wouldn't it be better to fire an error on the `onError` callback, if there is a failure to delete? Otherwise, the customer could be given the impression that the content was deleted locally and remotely, when perhaps neither is true.

**Comment by raphkim** [comment]
> Also, wouldn't it be better to fire an error on the onError callback, if there is a failure to delete? Otherwise, the customer could be given the impression that the content was deleted locally and remotely, when perhaps neither is true.

The items that customer directly ordered the deletion of _will_ be guaranteed to be deleted both locally and remotely if they receive the onResult callback. However, this implementation does not guarantee that all of the cascaded deletes were remotely sent. 
The philosophy of giving "best effort" to make sure that there is no orphaned items in cloud, but being "okay" if there are some is described [here](https://github.com/aws-amplify/amplify-android/pull/1058).

> Could you also add a unit test for the scenario that this fixes?

There is currently no specific issue that this PR addresses directly. Because we are "okay" with leaving orphaned items in remote storage, we simply decide to ignore any unexpected error that results while computing model relationship.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelTree.java [hunk 1]
```diff
@@ -140,9 +141,9 @@ private void recurseTree(
                                 childrenIds.add(cursor.getString(index));
                             } while (cursor.moveToNext());
                         }
-                    } catch (DataStoreException exception) {
+                    } catch (DataStoreException | SQLiteException exception) {
```
Comment (richardmcclellan) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelTree.java:144]: Should this be a more specific SQLite foreign key constraint exception?

**Comment by richardmcclellan** [comment]
> There is currently no specific issue that this PR addresses directly. Because we are "okay" with leaving orphaned items in remote storage, we simply decide to ignore any unexpected error that results while computing model relationship.

Now I'm a little confused.  If SQL's `ON DELETE CASCADE` is smart enough to identify the items that need to be deleted locally, why would we not be able to identify them in a similar way for the remote deletion?

File: aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelTree.java (hunk 1 above)
Comment (raphkim) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelTree.java:144]: This aims to ignore _any_ unexpected exception that can be thrown here, so if anything I would rather catch generic `Exception` here, but that would be code smell...
For example:
* Any SQLite specific behavior that was unexpected
  * 1000 variable/expression limit that we didn't know about would be an example of this (#1121 aims to fix this explicitly though)
* Incorrectly formatted graphQL schema that generated malformed models
  * #1146 would be an example of this. If customers started off with malformed models that happened to work out, we don't want to suddenly break their code by catching them here

**Comment by raphkim** [comment]
> Now I'm a little confused. If SQL's `ON DELETE CASCADE` is smart enough to identify the items that need to be deleted locally, why would we not be able to identify them in a similar way for the remote deletion?

SQLite does seem to be pretty efficient at discovering all related models... but I wasn't able to find a way to extract this information

**Comment by richardmcclellan** [comment]
> > Could you also add a unit test for the scenario that this fixes?
> 
> There is currently no specific issue that this PR addresses directly. Because we are "okay" with leaving orphaned items in remote storage, we simply decide to ignore any unexpected error that results while computing model relationship.


Doesn't this PR specifically address https://github.com/aws-amplify/amplify-android/issues/1146?  Couldn't we add a test case for it?

**Comment by raphkim** [comment]
> > > Could you also add a unit test for the scenario that this fixes?
> > 
> > 
> > There is currently no specific issue that this PR addresses directly. Because we are "okay" with leaving orphaned items in remote storage, we simply decide to ignore any unexpected error that results while computing model relationship.
> 
> Doesn't this PR specifically address #1146? Couldn't we add a test case for it?

This PR is not a direct solution to #1146. I do not think that a test case to see if deliberately breaking the schema will not crash cascading delete does not provide much value.

**Review by richardmcclellan — APPROVED** [review]
🚢  Approved, under the condition that we follow up by file a codegen bug to automatically generate bi-directional models, even if the customer provided a uni-directional schema.

**Comment by raphkim** [comment]
https://github.com/aws-amplify/amplify-codegen/issues/102
Addressing above codegen error will eliminate a major error case that this PR "ignores"

**Pull request merged**